src/main/java/com/google/devtools/build/lib/remote/RemoteSpawnRunner.java

```diff
@@ -699,6 +697,22 @@ private SpawnResult handleError(
         .build();
   }
 
+  static List<Property> getExtraPlatformProperties(Spawn spawn, Path execRoot) {
+    // Hack on Windows. The included headers dumped by cl.exe in stdout contain absolute paths.
+    // When compiling the file from different workspace, the shared cache will cause header
+    // dependency checking to fail. This was initially fixed by a hack (see
+    // https://github.com/bazelbuild/bazel/issues/9172 for more details), but is broken again due
+    // to cl/356735700. We include workspace name here so action results from different workspaces
+    // on Windows won't be shared.
+    boolean isWindows = System.getProperty("os.name").startsWith("Windows");
```

Comment:
This is the operating system Bazel runs on and that isn't necessarily the same as the remote operating system. In addition, this AFAIU needlessly decreases the cache hit rate for actions that do not invoke MSVC.

hat said, I don't know off the bat how to tell what the remote operating system is. I'm wondering if it's better to fix this by looking at the `PARSE_SHOWINCLUDES` feature in the C++ compile action. It should be possible to add the necessary property to the Spawn constructed in `CppCompileAction.createSpawn()`, which has the additional advantage that the hack stays in C++ code which requires it.

Reply:
> This is the operating system Bazel runs on and that isn't necessarily the same as the remote operating system.

This is used for remote cache only (which means local execution). For executing cl.exe remotely, RBE has the hack to remove the workspace name on the server side. And this requires other server implementation to do the same hack if they want to support compiling c++.

I had a discussion with @meteorcloudy yesterday about a long term fix for this. It seems we can solve the issue at Bazel side, but requires lots of changes. I will share the details later.

> In addition, this AFAIU needlessly decreases the cache hit rate for actions that do not invoke MSVC.

Note that, if we use parent of exec root as input root, the paths of input files include workspace name (input file paths are used to compute cache key) which implies the same thing: caches across different workspace won't be shared. So I won't concern too much about this.

> I'm wondering if it's better to fix this by looking at the PARSE_SHOWINCLUDES feature in the C++ compile action.

Good point! I will have a look.